Prevent `</p>` output in Rich Text control

kienstra opened the pull request:
## Steps to reproduce
1. Create a new post
2. Add a Block Lab block that has a Rich Text field without a default value
3. Don't enter any text in the Rich Text field
4. Publish the post, and go to the front-end
5. Expected: The output is well-formed html
6. Actual: The output is `</p>`, as reported in #437:

<img width="528" alt="closing-p" src="https://user-images.githubusercontent.com/4063887/66259261-7f54d780-e774-11e9-8c71-8a0ccaac91fe.png">

It looks like when the <RichText> component has no text, it saves `<p></p>`. But passing that to wpautop() returns `</p>`:

<img width="982" alt="wpautop-here" src="https://user-images.githubusercontent.com/4063887/66259292-d9559d00-e774-11e9-91b0-2d6385195320.png">

So in that case, this PR simply returns the string.

Closes #437

kienstra committed:
Prevent </p> output in Rich Text control

When the <RichText> component has no text,
it saves '<p></p>'.
But passing that to wpautop() returns '</p>'.
So in that case, simply return the string.

kienstra committed:
Merge branch 'develop' into fix/rich-text-empty

kienstra committed:
Fix failing unit tests, seemingly not related to this PR

It looks like I didn't use date() correctly in the tests.
The first parameter should be the format,
and I had passed it '+1 month'.
I'm not sure how it worked before.

kienstra commented:
**Request For Review**

Hi @lukecarbis,
When you have a chance, could you please review this? 

Thanks!

lukecarbis commented on php/blocks/controls/class-rich-text.php:
```diff
@@ -51,6 +51,12 @@ public function register_settings() {
 		unset( $echo );
+
+		// For this value, wpautop() returns '</p>', so simply return the value.
+		if ( '<p></p>' === $value ) {
```
Something like this might handle more weird edge cases (such as `<span></span>` or `<p><strong></strong></p>`):

```PHP
if ( empty( esc_html( $value ) ) ) {
    return $value;
}
// or…
if ( '' === esc_html( $value ) ) {
    return $value;
}
// or even…
if ( '' === esc_html( $value ) ) {
    return '';
}
```

kienstra replied:
Ah, interesting point.

I'm not sure if there will be a case where `empty( esc_html( $value ) )`, or `'' === esc_html( $value )`, though maybe I just haven't seen it. 

It looks like even if there's no value entered in the Rich Text control, its value is `<p></p>`

lukecarbis replied:
I could be missing something here.
- If the Rich Text control value is empty, the value is set to `<p></p>`
- The expected value would be an empty string, right?
- The logic above says "After removing all the HTML, if I'm left with an empty string, return an empty string"

kienstra replied:
Hi @lukecarbis,
All good, those are good points.

>If the Rich Text control value is empty, the value is set to `<p></p>`

Right 😄 

>The expected value would be an empty string, right?

Good question, maybe 😄 

The way I thought of this was like the Paragraph block, which also uses the `<RichText>` component. It outputs `<p></p>` when it's empty, just like this Rich Text control does:

<img width="988" alt="empty-paragraph-block" src="https://user-images.githubusercontent.com/4063887/66356531-db645b00-e930-11e9-80de-d5a002f8af93.png">

This is definitely debatable, though. The paragraph block isn't exactly like this.


>The logic above says "After removing all the HTML, if I'm left with an empty string, return an empty string"

Ah, `esc_html()` doesn't usually remove HTML, it just escapes it. 

For example, `esc_html( '<div></div>' )` returns `&lt;div&gt;&lt;/div&gt;`

kienstra replied:
Maybe [strip_tags()](https://www.php.net/manual/en/function.strip-tags.php)

lukecarbis replied:
Ah – of course! Sorry, I did mean strip_tags. Even better, though: [`wp_strip_all_tags`](https://developer.wordpress.org/reference/functions/wp_strip_all_tags/).

kienstra replied:
Sure, I thought that's what you meant.

kienstra replied:
Though it's unlikely, it's possible that a Rich Text field has only an image, in which case it would be stripped entirely, as there's no text other than the markup:

<img width="855" alt="image-in-rich-text" src="https://user-images.githubusercontent.com/4063887/66361071-00fa6000-e943-11e9-9191-0a67cf4239c3.png">

kienstra replied:
https://github.com/getblocklab/block-lab/pull/444/commits/a65299c07cbe1432f27b760aa511c2dcc12b4ef6 changes this to return `''` if the Rich Text field is empty (if it saves `<p></p>`).

lukecarbis replied:
Aah! Great pickup on the image!

kienstra replied:
Thanks, it's kind of an edge case. Otherwise, `strip_tags()` would work really well.

lukecarbis reviewed (approved):
One small suggestion here, feel free to merge with or without it.

kienstra committed:
If the Rich Text field has '<p></p>', output ''

This is the behavior of the Text and Classic Text
fields.
This is debatable, but it keeps the fields consistent.

kienstra commented:
Hi @lukecarbis,
Happy weekend!

What do you think of the [latest changes](https://github.com/getblocklab/block-lab/pull/444#discussion_r332307316) here?

Thanks!

lukecarbis commented:
Great! Going to merge this on you behalf, if you don't mind, so we can get that unit test fix in a couple of other PRs.

kienstra commented:
Sounds good, thanks for merging this.
